### ISSUE TITLE
Documentation on using Cue and Kubernetes

### ISSUE THREAD

**Issue by jmgilman** [stefanprodan/timoni#251]
First of all, thanks for this interesting project, we're currently evaluating using it internally at our company. That being said, the documentation is great for getting started with Timoni, but for users new to CUE + Kubernetes, there doesn't seem to be much for getting started. In other words, I could read both the Timoni and CUE docs and still have very few examples of how the Timoni docs came up with some of the given examples. 

Are there resources out there that I am missing?

**Comment by stefanprodan** [comment]
There is nothing specific in CUE for Kubernetes, Timoni comes with Kubernetes schemas that you can use to do resource templating, which work like any other CUE value. Please see https://timoni.sh/module/ and let me know what things you would like to see in there specific to Kubernetes. I suggest starting with `timoni mod init app .` and trying to add a new resource like in the docs.

**Comment by jmgilman** [comment]
```
❯ timoni mod init myapp .
4:02PM INF module initialized at myapp
❯ cd myapp                                                                      
❯ cue cmd -t debug -t name=test -t namespace=test -t mv=1.0.0 -t kv=1.28.0 build
import failed: expected '}', found 'EOF':
    ./timoni.cue:11:2
    ./templates/config.cue:68:3
❯ cue version                                                                   
cue version 0.5.0

go version go1.20.8
      -buildmode exe
       -compiler gc
       -trimpath true
     CGO_ENABLED 1
          GOARCH amd64
            GOOS linux
         GOAMD64 v1
```

Unfortunately, I can't get linting to work, but I don't see anything immediately wrong with the specified file.

**Comment by jmgilman** [comment]
Just a few initial comments from a complete beginner:

The result of `mod init` is _very_ large. I understand the value of a deployable example, but there's so much wiring going on that it took a while for me to understand how things were being compiled together. In addition to that, because things are so tightly wired, removing some of the example resources ended up breaking things in multiple places. For example, what if I don't want a test job? I was surprised at how much I had to change when I deleted the job file. 

Perhaps an additional flag to `mod init` that adds the complete example would be more appropriate. Something like the following seems to be the minimal amount needed:

- cue.mod
- templates
   - templates/config.cue
- debug_*.cue
- timoni.cue
- timoni.ignore
- values.cue

As it stands, for me to "get started" with my first project, I had to spend quite a bit of time unwiring the starting template.

**Comment by stefanprodan** [comment]
> I realized there's a newer version of cue which fixed the issue (the NixOS package is behind).

Timoni embeds the CUE engine, all commands will work even without the cue CLI installed. But yes, you need CUE 0.6 to play with Timoni modules.

> As it stands, for me to "get started" with my first project, I had to spend quite a bit of time unwiring the starting template.

Lots of people new to CUE said the same thing. I agree that we need a module example that outputs a single Kubernetes object, maybe a ServiceAccount would do. The current module used for init is very much like what you get at `helm create`, an nginx chart that has all things you would need to deploy a read app.

**Comment by jmgilman** [comment]
> Lots of people new to CUE said the same thing.

Yeah, it's a bit unfair to pin the initial complexity of CUE on Timoni, but it's probably likely to happen quite often given people from the Kubernetes community will likely find Timoni before CUE (or Timoni finally gets them interested in it like in my case). So Timoni is in an interesting spot where it's utilizing a lot of CUE powers (modules, commands, schemas, defaults etc.) in the helm-style "Hello, world" example, but all combined it's a pretty overwhelming experience for someone totally new to both projects. 

I think a more simple init solves two problems, really:

1. It's a bit less overwhelming as I've previously described.
2. I've already found a need for this simplified version because I need to package a dozen or so microservices and copy/pasting the minimal template is a lot easier than unwiring the more complete example.

As for documentation improvements, I might have some ideas, but I'm still working on them. I took some notes today that I'll process further. Overall, it took a couple of hours before I could understand the full flow between the entry point, templates, external packages, etc. Which isn't terrible in the long run, but it certainly took some resolve to push through :) 

On the good side of things, this project is a huge improvement over Helm and solves almost all of the issues I have with it. So I'm very excited to see it grow and contribute where I can.

**Comment by stefanprodan** [comment]
For the docs website, I think we need a full-blown section for "Module development" that could start with the basics of CUE lang and CUE modules, followed by Timoni module structure and the machinery put in place with Kubernetes schemas, templates, config, entry point, etc. It could also contain advanced topics like, how do you generate immutable ConfigMaps/Secrets and trigger a Deployment rolling upgrade (this is showcased in the init module without any docs), how do you write e2e tests with Kube Jobs, multi-stage apply (what the Redis module shows, with master/replicas orchestration).

**Comment by stefanprodan** [comment]
> I've already found a need for this simplified version because I need to package a dozen or so microservices and copy/pasting the minimal template is a lot easier than unwiring the more complete example.

One way to solve this, would be for init to have a `--blueprint-url` flag where you could set the OCI URL to some other module then default one. Timoni would perform the replacements and generate the module locally like it does now. We just need to decide what keyword to replace, for now it's hardcoded to `minimal` https://github.com/stefanprodan/timoni/blob/59e75c1cb4f707b2778b7bddbae9240cb4bde578/cmd/timoni/mod_init.go#L57

**Comment by stefanprodan** [comment]
Hey @jmgilman this is great advice thank you. I will start with commenting the Config and if you could help review that would be great. I'm also considering having a more lightweight blueprint. As for enforcing the image digest, module authors can simply disable that by setting `digest: ""` in the `values.cue`, this is for the test image in current module blueprint.

**Comment by jmgilman** [comment]
This is what I am currently using as a template: https://gist.github.com/jmgilman/641857c212e04dc18cc444051bfe5046

**Comment by stefanprodan** [comment]
@jmgilman if you have time, please review #261. Thanks

**Comment by jmgilman** [comment]
Done :)

**Comment by jmgilman** [comment]
That looks great! A user can start there and then graduate to a more complicated example which introduces more of the "bells and whistles" of CUE/Timoni :)
